docs/RFCS/20200414_scheduled_jobs.md

```diff
@@ -0,0 +1,383 @@
+
+The above steps run inside a single transaction to ensure that
+only one node starts running a job. Any errors encountered during
+execution are handled based on the `ScheduleDetails` configuration.
+
+The important thing to keep in mind is that since we plan on executing
+inside the transaction (at least for the initial version), the actual
+statement must complete fairly quickly.  To that end, we plan on 
+modifying relevant statements (`BACKUP`, `IMPORT`, etc) so that these
+statements do not block for the duration of the operation, but instead
+complete upon the creation of the job record.
+
+#### Job completion hook
+Recall, the scheduled job framework is responsible for spawning the
+system job to do the actual work.  The scheduler needs to know
+about the status of the completed jobs.
+
+Depending on the schedule configuration, failed runs may pause the cron job.
+The scheduled job may also want to persist some state based on the results
+of the job (e.g. save the last full backup location in a table)
+
+We will also modify the jobs system to inform scheduled jobs whenever the scheduled
+job terminates:
+
+```go
+func NotifyJobCompletion(ctx, md *JobMetadata, txn *Txn) error
+
+``` 
+
+The `jobspb.JobMetadata` contains all the relevant job information:
+job ID, schedule ID, job completion status, as well as opaque payload,
+describing the job itself.  The notification hook is also given a
+transaction, so that the changes performed by the schedule notification
+hook happens atomically with the job status update mutation.
+
+
+#### Polling
+Our initial MVP version will use a simple polling mechanism where each node picks 
+the next runnable job, and attempts to execute it.
+
+Inside the transaction, we select the next eligible job to run.
+
+```sql
+SELECT 
+  C.job, C.schedule, C.exec_spec, C.user,
+  (SELECT count(*) 
+   FROM system.jobs J
+   WHERE J.sched_id = C.sched_id AND job_status NOT IN ('failed', 'succeeded')
+  ) AS num_running,
+FROM system.crontab C
+WHERE next_run < current_timestamp()
+```
+
+If `num_running > 0` and policy setting `skip_wait == false`, then we need to
+wait for the previous invocation to complete.  To do this, we simply advance
+the `next_run` time by some amount and commit transaction.
+
+Once job execution policy constraints have been satisfied, 
+we can compute the next time this job is supposed to run.
+If the schedule is set, then compute the next_run based on the schedule.
+If not, this is a one-off job, set `next_run` to `NULL` (and update
+`change_info` to indicate that the schedule completed).
+
+Finally, execute the statement and commit the transaction.
+
+We have multiple venues on improving scalability and performance in the future releases,
+driven by the customer needs (Cockroach Cloud, etc): 
+  * Nodes may pick a random runnable job (instead of the oldest), to reduce contention
+  * If polling becomes problematic (scale), it can be replaced by sharding the jobs to run
```

Comment:
I'd just say we can explore using another mechanism (leasing/locks) to coordinate which node has exclusive dispatch responsibility, and then allow that node to load the schedule into a more efficient structure like a timer ring that is populated via polling or a rangefeed.